### ISSUE TITLE
Disable this damn : "should have comment or be unexported"

### ISSUE THREAD

**Issue by okabbas** [golang/lint#389]
Please delete the message **should have comment or be unexported** .
This message will hurt my nerve in **vscode** and I'm not endearing these messages in my code yet.

Whenever necessary, I add to my comment code and do not need to display this message.
What is the reason why you decide for me and make sure you comment on the code?

**Comment by andybons** [comment]
Thanks for the issue.

Unfortunately, this check will remain intact since it's within the [scope](https://github.com/golang/lint/blob/master/README.md#scope) of the tool. CodeReviewComments states:

> All top-level, exported names should have doc comments, as should non-trivial unexported type or function declarations. See https://golang.org/doc/effective_go.html#commentary for more information about commentary conventions.

Also, from the readme:
> The suggestions made by golint are exactly that: suggestions. Golint is not perfect, and has both false positives and false negatives. Do not treat its output as a gold standard. We will not be adding pragmas or other knobs to suppress specific warnings, so do not expect or require code to be completely "lint-free". In short, this tool is not, and will never be, trustworthy enough for its suggestions to be enforced automatically, for example as part of a build process.

I'm sorry this is getting on your nerves. You can use gometalinter in place of golint which lets you disable certain checks as I understand it.

**Comment by jackson-sandland** [comment]
The issue is that most of the time if you are writing small functions, functions names will be trivial. The tool scope says that `non-trivial unexported type or function declarations` should have doc comments. This should not be default behavior.

**Comment by jirfag** [comment]
golangci-lint disables this annoying warning by default

**Comment by upitau** [comment]
One of good software engineering practices is to produce self-documenting code. Which means that the code should be written in such a way, so that documentation becomes unnecessary (in most cases) as it wouldn't add any useful information. This linting rule ruins the practice.

**Comment by andriykutsevol** [comment]
You are like a child. Have you ever worked on a limited budget and time?
Most people work this way in the field of IT today.
No one is interested in all these lyrics.

**Comment by ianlancetaylor** [comment]
@andriykutsevol Please stay polite.  Please follow the Code of Conduct at https://golang.org/conduct.  Thanks.

**Comment by andriykutsevol** [comment]
Have you ever worked on a limited budget and time?
Most people work on a limited budget and time these days.
No one is interested in all these rules.

**Comment by andriykutsevol** [comment]
Most people work on a limited budget and time these days.
No one is interested in these rules.

**Comment by sundbergad** [comment]
I am new to golang coding but have as my first issue found this to be very disturbing.
In my case even with flaws. This is the mess I got... so what is wrong with my exported function name ```
func SizeOfDatatype(){ 
    ... implementation ....
 }
```
----
 comment on exported function SizeOfDatatype should be of the form "SizeOfDatatype ..." 
 --- 

To me this makes no sence and as many of you guys mention, make it ignorable at least ... Please

**Comment by szank** [comment]
I feel like I need to strongly oppose the flood of negative remarks about golint in this thread.

In my opinion golint, enforced exported method comments and godoc is a boon for the language and for the developers. 

Forced documentation means that 
1. You need to think about what to export. If function/type/const need not to be exported, you don't need to write documentation for it. It's a good incentive to export only what's necessary.

2. Ease of adding documentation to the code and the common docs presentation format (godoc) is one of the reasons why I consider go's open source libraries the best documented libraries in any language I have seen. 

3. Ease of adding documentation allows people to create PRs to fix just documentation in the open source library they are using. Which is a great thing. 

4. In my team the golint have created an environment where we held each other accountable for maintaining proper documentation for our internal code. I guess this would happen anyway, but it's easier to do it when you have external tools enforcing this. 

Still, documentation of some go's open source libraries leave me wanting, but in general I am very very grateful for golint and godoc. 

It makes my work much, much easier.

**Comment by sundbergad** [comment]
szank, it may be that you and your team need it, however my team and I don't need it and with the flaws in the warning that I've commented above, we don't see why it isn't an option to enable/disable as a minimum.

**Comment by andriykutsevol** [comment]
If someone so serious about this things let them do it by themselves.
It is very annoying behavior.

**Comment by ianlancetaylor** [comment]
The lint program is intended to encourage people to write code in a specific style, as described at https://golang.org/wiki/CodeReviewComments with reference to https://golang.org/doc/effective_go.html.  If you do not want to write code in that style, then do not run the lint program.  The lint program is not supposed to be for all people for all uses; that's what the vet program is for.

**Comment by andriykutsevol** [comment]
"The lint program is not supposed to be for all people for all uses"
So it should be disabled by default?

Today's average software developer have to keep an eye for enormous different frameworks, languages and libraries.

And when we see some "paranoidal" rule like this enabled by default, It makes us feel very bad.

**Comment by andriykutsevol** [comment]
Thank you.

**Comment by infografx** [comment]
When I want to quick write the code, without much worried about the code style, I  use the next settings.json (Visual Studio Code):

{
    "workbench.colorTheme": "Visual Studio Light",
    "go.buildOnSave": "off",
    "go.lintOnSave": "off",
    "go.vetOnSave": "off",
    "go.buildFlags": [],
    "go.vetFlags": [],
    "go.useCodeSnippetsOnFunctionSuggest": false,
    "go.formatTool": "goreturns",
    "go.lintTool": "gometalinter",
    "go.lintFlags": ["--disable=all"],
    "go.formatFlags": [
        "--disable=all"
    ],
    "python.jediEnabled": false,
    "python.linting.pylintEnabled": false,
    "python.autoComplete.extraPaths": [
        "/some/path"
    ]
}

**Comment by patrikeh** [comment]
@szank have you seen packages that try to enforce this to meet a decent go report rating or to silence these warnings? They are riddled with useless comments like

// GetUser retrieves a user

**Comment by andriykutsevol** [comment]
In general I respect USA, Google, etc... 

But.

Is this the use case?

https://edition.cnn.com/world/live-news/boeing-737-max-8-ethiopia-airlines-crash/h_9dee508ccc0a80f533a2bdafb8d5b536

```
That suggests the plane was not responding to pilots’ inputs and they were losing control of the aircraft, Quest explains.
```

**Comment by Hyvi** [comment]
> golangci-lint disables this annoying warning by default

golangci-lint is a better choice, I start to use golangci-lint in CI within gitlab

**Comment by cmitchellangelo** [comment]
I just removed the capital letter and it went away :)

**Comment by ms-xy** [comment]
You know, I do understand the basic idea and the benefits. However, in my opinion developers got to be realistic instead of idealistic.

- If I as a developer feel encumbered by needless warnings, I turn them off
- If I cannot turn off a specific warning, I turn them all off
- Thus I might miss an actually valid warning

or (in this special case):

- I write bullshit / placeholder comments to satisfy the linter

Go! figure.

**Comment by elgs** [comment]
Problem solved by:
```
rm -rf $HOME/go/bin/golint
```
Usually I don't like to do such extreme things, but golint is way too extreme than I could stand.

**Comment by ilius** [comment]
Adding a flag does not contradict with Effective Go or the standard practice.
Python also has [PEP8](https://www.python.org/dev/peps/pep-0008/), but `pep8` allows you to switch off/on error/warning codes.
Being minimalistic about core language feature is good, but this is just a linting tool!

**Comment by devinbost** [comment]
One of the maintainers here needs to look at the number of up-votes and number of down-votes to the various comments and tell me what the community really wants.

**Comment by devinbost** [comment]
@andybons One of the maintainers here needs to look at the number of up-votes and number of down-votes to the various comments and tell me what the community really wants.

**Comment by elgs** [comment]
I feel two decades ago, people tended to do things like, this is a reference implementation of an idea, and it's free for the market to provide better implementations. Most recently, people tended to use some thing called lint to force others to accept their opinionated opinion, otherwise, they would give others some hard time like errors/warnings.

**Comment by ilius** [comment]
Another benefit of allowing to select error codes (like `pep8`) is that it allows you to prioritize fixing them, some of them can be mandatory for ci to pass, some can be shown as warnings in pull requests, some can be checked and fixed periodically (or before releases), some of them can be ignored completely...

**Comment by andybons** [comment]
Thanks, everyone.

For now, we will keep this check unchanged. As [documented](https://github.com/golang/lint#scope) This tool is meant to mechanically enforce rules laid out by [Effective Go](https://golang.org/doc/effective_go.html) and [CodeReviewComments](https://golang.org/wiki/CodeReviewComments), with this rule specifically enforcing [comment sentences](https://golang.org/wiki/CodeReviewComments#comment-sentences).

Additionally, the README states:

> The suggestions made by golint are exactly that: suggestions. Golint is not perfect, and has both false positives and false negatives. Do not treat its output as a gold standard. We will not be adding pragmas or other knobs to suppress specific warnings, so do not expect or require code to be completely "lint-free". In short, this tool is not, and will never be, trustworthy enough for its suggestions to be enforced automatically, for example as part of a build process.

Changing this would mean changing Golint’s scope, Effective Go, and/or CodeReviewComments. There are appropriate forums to discuss those changes, but this issue is not one of them.

Thanks
